### PR TITLE
Implement navigation to test source for tests defined in the test file

### PR DESCRIPTION
**Pull request by stevemessick**
@alexander-doroshko You can double-click a test result, or use the context menu, to show the test source code in the editor. This only works for tests define in the file loaded by the test runner. According to package:test maintainers that's how tests should be written -- there is no need for something like all_tests.dart anymore. We will have to talk about the desired UX for IntelliJ.

### CODE REVIEW

File: Dart/src/com/jetbrains/lang/dart/ide/runner/test/DartTestEventsConverter.java [hunk 1]
```diff
@@ -151,6 +166,11 @@ private boolean handleTestStart(JsonObject obj) throws ParseException {
     return result;
   }
 
+  private static String encodeLocation(String url, Item item) {
```
Comment (alexander-doroshko) [inline Dart/src/com/jetbrains/lang/dart/ide/runner/test/DartTestEventsConverter.java]: Method is unused. Can it be removed?

**Comment by alexander-doroshko** [comment]
`DartTestLocationProviderTest` has to be updated.

**Comment by alexander-doroshko** [comment]
Will it make sense to handle that theoretically can be thrown by `fromJson()` methods used in different places?

**Comment by alexander-doroshko** [comment]
Review completed. Need to update broken test before push.

**Comment by alexander-doroshko** [comment]
Awesome side effect of the fix is that now it is possible to run a single test by a right-click on the node in test result tree!

Related question: is it possible to implement 'Jump to source' for test group nodes as well?

File: Dart/src/com/jetbrains/lang/dart/ide/runner/test/DartTestEventsConverter.java (hunk 1 above)
Comment (stevemessick) [inline Dart/src/com/jetbrains/lang/dart/ide/runner/test/DartTestEventsConverter.java]: Yes. I forgot to delete it after inlining it.

**Comment by stevemessick** [comment]
*Will it make sense to handle that theoretically can be thrown by fromJson() methods used in different places?*

You mean doing something other than what processEventText() does? I don't really think it is necessary. That one catches all event-parsing exceptions. Decoding the URL in DartTestLocationProvider should not throw any exceptions since it is decoding JSON produced by the event converter.

*Awesome side effect of the fix is that now it is possible to run a single test by a right-click on the node in test result tree!*

How do you get that? I don't see it.

*Related question: is it possible to implement 'Jump to source' for test group nodes as well?*

Good question. I don't know how that is implemented.

**Comment by alexander-doroshko** [comment]
Don't you have it in context menu?
image

Mechanism is the following: 
- `DartTestLocationProvider` returns `PsiElement` that corresponds to the node in test result tree. Returned `PsiElement` is a `DartCallExpression` that corresponds to `test()` or `group()` method call.
- `DartTestRunConfigurationProducer` suggests run configuration based on context `PsiElement`. For  `test()` or `group()` calls it suggests to sun a single test/group correspondingly. It is the same as if you right click on a `test()` or `group()` call in editor.

**Comment by stevemessick** [comment]
My context menu is different. The Run item says it will run the whole test
suite again, not an individual test, even though I have a single test
selected. I'm not sure why it is different from yours.

As to groups (since I'm not sure which you were asking about) I don't get
much of a context menu for them. It does not have the run or navigate
items. Again, I'm not sure why.

Both of these things need to be fixed IMO. But I think they can wait while
I figure out how to get test navigation working, since that may require
changes to the test runner.

**Pull request closed without merging**